File: eseller_suite/eseller_suite/custom_script/purchase_invoice/purchase_invoice.py
# Copyright (c) 2025, efeone and contributors
# For license information, please see license.txt

import frappe

def before_save(doc, method):
	'''
		Method which trigger on before_save event of Purchase Invoice
	'''
	set_bundle_diff_amount(doc)
	set_discount_based_on_amazon_value(doc)

def on_submit(doc, method):
	'''
		Method which trigger on on_submit event of Purchase Invoice
	'''
	if doc.bundle_difference_amount >=1 or doc.bundle_difference_amount <= -1:
		title = 'Check Difference Amount'
		msg = 'Cannot submit the invoice due to difference amount of {0} for bundle items'.format(frappe.bold(doc.bundle_difference_amount))
		frappe.throw(
			title=title,
			msg=msg
		)
	if doc.amazon_invoice_id:
		unset_stn_exception(doc)

def on_update_after_submit(doc, method):
	'''
		Method which trigger on on_update_after_submit event of Purchase Invoice
	'''
	unset_stn_exception(doc)

def on_cancel(doc, method):
	'''
		Method which trigger on on_cancel event of Purchase Invoice
	'''
	doc.ignore_linked_doctypes = (
		"GL Entry",
		"Stock Ledger Entry",
		"Repost Item Valuation",
		"Repost Payment Ledger",
		"Repost Payment Ledger Items",
		"Repost Accounting Ledger",
		"Repost Accounting Ledger Items",
		"Unreconcile Payment",
		"Unreconcile Payment Entries",
		"Payment Ledger Entry",
		"Tax Withheld Vouchers",
		"Serial and Batch Bundle",
		"Amazon STN Entry",
	)

def set_bundle_diff_amount(doc):
	'''
		Method to set bundle differences and total
	'''
	#Calculating based on Items table
	total_bundle_amount = 0
	for row in doc.items:
		if row.from_bundle_item:
			total_bundle_amount += row.amount
	doc.total_bundle_amount = round(total_bundle_amount, 2)

	#Calculating based on Bundle Items Table
	total_bundle_amount_actual = 0
	for item in doc.bundle_items:
		if item.rate and item.qty:
			item.amount = item.rate * item.qty
		total_bundle_amount_actual += item.amount
	doc.total_bundle_amount_actual = round(total_bundle_amount_actual, 2)

	# Setting Difference Amount
	doc.bundle_difference_amount = round((doc.total_bundle_amount_actual - doc.total_bundle_amount), 2)

def set_discount_based_on_amazon_value(doc):
	'''
		Method to set dicount based on Amazon value and outstanding amount
	'''
	if doc.amazon_invoice_value and doc.outstanding_amount:
		diff = doc.outstanding_amount - doc.amazon_invoice_value
		# Add discount if difference is between -1 to 1, else it may be some error
		if diff and (1 > diff > -1):
			doc.apply_discount_on = 'Grand Total'
			doc.discount_amount = diff

def unset_stn_exception(doc):
	'''
		Method to unset STN exception after invoice submission
	'''
	if frappe.db.exists('Amazon STN Entry Item', {'purchase_invoice': doc.name}):
		stn_entry_item, si_ref  = frappe.db.get_value('Amazon STN Entry Item', {'purchase_invoice': doc.name}, ['name', 'sales_invoice'])
		frappe.db.set_value('Amazon STN Entry Item', stn_entry_item, 'error_log', '')
		frappe.db.commit()
		if frappe.db.get_value('Sales Invoice', si_ref, 'docstatus') == 0:
			try:
				si_doc = frappe.get_doc('Sales Invoice', si_ref)
				si_doc.submit()
			except Exception as e:
				frappe.db.rollback()
				exception_msg = f"Failed to submit Sales Invoice: {si_ref} - {str(e)}"
				frappe.db.set_value('Amazon STN Entry Item', stn_entry_item, 'error_log', exception_msg, update_modified=False)

File: eseller_suite/eseller_suite/custom_script/purchase_invoice/purchase_invoice.js
// Copyright (c) 2025, efeone and contributors
// For license information, please see license.txt

frappe.ui.form.on("Purchase Invoice", {
	refresh(frm) {
		frm.set_query("item_code", "bundle_items", () => {
			return {
				filters: {
					is_bundle_item: 1
				}
			};
		});
		manage_custom_buttons(frm);
	}
});

frappe.ui.form.on('Purchase Invoice Item', {
	item_code: function (frm, cdt, cdn) {
		apply_bundle_items(frm, cdt, cdn);
	},
	qty: function (frm, cdt, cdn) {
		apply_bundle_items(frm, cdt, cdn);
	},
	bundle_items_remove: function (frm, cdt, cdn) {
		let deleted_row_name = cdn;
		(frm.doc.items || []).forEach(d => {
			if (d.bundle_parent === deleted_row_name) {
				frappe.model.clear_doc(d.doctype, d.name);
			}
		});
		frm.refresh_field("items");
	},
});

/*
 * When an item is selected in the bundle_items child table, fetch its bundle components and add them as child rows.
*/
function apply_bundle_items(frm, cdt, cdn) {
	let row = locals[cdt][cdn];
	if (row.parentfield !== "bundle_items") return;
	if (!row.item_code) return;

	frappe.call({
		method: "eseller_suite.eseller_suite.utils.get_bundle_items",
		args: {
			bundle_item: row.item_code
		},
		callback: function (r) {
			if (!r.message) return;

			let selected_qty = flt(row.qty) || 1;

			(frm.doc.items || []).forEach(d => {
				if (!d.item_code) {
					frappe.model.clear_doc(d.doctype, d.name);
				}
			});

			(frm.doc.items || []).forEach(d => {
				if (d.bundle_parent === row.name) {
					frappe.model.clear_doc(d.doctype, d.name);
				}
			});

			r.message.forEach(function (bundle_child) {
				let new_row = frm.add_child("items");
				new_row.item_code = bundle_child.item_code;
				new_row.item_name = bundle_child.item_name;
				new_row.qty = flt(bundle_child.qty) * selected_qty;
				new_row.uom = bundle_child.uom;
				new_row.rate = bundle_child.rate;
				new_row.description = bundle_child.description;
				new_row.bundle_parent = row.name;
				new_row.bundle_qty = bundle_child.qty;
				new_row.from_bundle_item = 1;
			});
			frm.refresh_field("items");
		}
	});
}

function manage_custom_buttons(frm) {
	if (frm.doc.amazon_invoice_id) {
			if (frm.doc.inter_company_invoice_reference) {
				frm.add_custom_button(__('Inter-Company Sales Invoice'), () => {
					frappe.set_route('Form', 'Sales Invoice', frm.doc.inter_company_invoice_reference);
				}, __('View'));
			}
	}
}
